src/mapper/pkg/resolvers/intents_store.go

```diff
@@ -0,0 +1,130 @@
+package resolvers
+
+import (
+	"encoding/json"
+	"github.com/otterize/network-mapper/src/mapper/pkg/graph/model"
+	"github.com/samber/lo"
+	"sort"
+	"time"
+)
+
+type ServiceToDestToTimestamp map[model.OtterizeServiceIdentity]map[model.OtterizeServiceIdentity]time.Time
+
+type DiscoveredIntent struct {
+	Source      model.OtterizeServiceIdentity
+	Destination model.OtterizeServiceIdentity
+	Timestamp   time.Time
+}
+
+type IntentsHolderStore struct {
+	serviceMap ServiceToDestToTimestamp
+}
+
+func NewIntentsHolderStore() IntentsHolderStore {
+	return IntentsHolderStore{
+		serviceMap: make(ServiceToDestToTimestamp),
+	}
+}
+
+func (serviceMap ServiceToDestToTimestamp) MarshalJSON() ([]byte, error) {
+	// OtterizeServiceIdentity cannot be serialized as map key in JSON, because it is represented as a map itself
+	// therefore, we serialize the store as a slice of [Key, Value] "tuples"
+	sourceToIntents := make(map[model.OtterizeServiceIdentity][]DiscoveredIntent)
+	for source, destinations := range serviceMap {
+		var intents []DiscoveredIntent
+		for destination, timestamp := range destinations {
+			intents = append(intents, DiscoveredIntent{
+				Source:      source,
+				Destination: destination,
+				Timestamp:   timestamp,
+			})
+		}
+		sourceToIntents[source] = intents
+	}
+	return json.Marshal(lo.ToPairs(sourceToIntents))
+}
+
+func (serviceMap ServiceToDestToTimestamp) UnmarshalJSON(b []byte) error {
+	var pairs []lo.Entry[model.OtterizeServiceIdentity, []DiscoveredIntent]
```

Comment:
Avoid using lo structs for serialization. They aren't meant for it, and while it compiles and works now, it can change with versions of lo. If you serialize, use your own structs - that way, you control how things get serialized, and you know that old data you serialized can be deserialized with the same struct.